### PR TITLE
Filter list response with unique results

### PR DESCRIPTION
**Pull request by andresmgot**
Signed-off-by: Andres Martinez Gotor <andres@bitnami.com>

Ref: #570

Return unique charts by digest when listing.

cc/ @prydonius

### CODE REVIEW

File: cmd/chartsvc/handler.go [hunk 1]
```diff
@@ -217,11 +217,17 @@ func newChartResponse(c *models.Chart) *apiResponse {
 }
 
 func newChartListResponse(charts []*models.Chart) apiListResponse {
+	chartDict := map[string]*apiResponse{}
```
Comment (prydonius) [inline cmd/chartsvc/handler.go]: I would call this `chartDigests` and add a comment above this to describe that this is used to track chart versions by digests to ensure uniqueness

File: cmd/chartsvc/handler.go [hunk 2]
```diff
@@ -217,11 +217,17 @@ func newChartResponse(c *models.Chart) *apiResponse {
...
 	cl := apiListResponse{}
 	for _, c := range charts {
-		cl = append(cl, newChartResponse(c))
+		chartResponse := newChartResponse(c)
+		// The same chart may be present in different repos, filter just one
+		id := (chartResponse.Relationships["latestChartVersion"].Data).(models.ChartVersion).Digest
```
Comment (prydonius) [inline cmd/chartsvc/handler.go]: this might be confusing because the `id` in the response and mongodb is `reponame/chartname`, not the digest, we could just call this variable `digest`

File: cmd/chartsvc/handler.go [hunk 3]
```diff
@@ -217,11 +217,17 @@ func newChartResponse(c *models.Chart) *apiResponse {
...
+		if found := chartDict[id]; found == nil {
```
Comment (prydonius) [inline cmd/chartsvc/handler.go]: whilst this does work, I think a more correct way of doing this would be:

```
if _, ok := chartDict[id]; !ok {
...
}
```

The use of `ok` instead of `found` is a Go idiom that I see regularly, so I think it's better to use that name.

File: cmd/chartsvc/handler.go (hunk 2 above)
Comment (prydonius) [inline cmd/chartsvc/handler.go]: wouldn't it be simpler to just do

```suggestion
		id := c.ChartVersions[0].Digest
```
That way we avoid having to assert the type.

We can use `newChartResponse` below when we append to `cl`.

File: cmd/chartsvc/handler_test.go [hunk 4]
```diff
@@ -126,28 +126,40 @@ func Test_newChartResponse(t *testing.T) {
...
+			{ID: "stable/wordpress", ChartVersions: []models.ChartVersion{{Version: "1.2.3", Digest: "1234"}, {Version: "1.2.2", Digest: "12345"}}},
+		}, []*models.Chart{
+			{ID: "my-repo/my-chart", ChartVersions: []models.ChartVersion{{Version: "0.0.1", Digest: "123"}}},
+			{ID: "stable/wordpress", ChartVersions: []models.ChartVersion{{Version: "1.2.3", Digest: "1234"}, {Version: "1.2.2", Digest: "12345"}}},
+		}},
+		{"has a duplicated chart", []*models.Chart{
+			{ID: "my-repo/my-chart", ChartVersions: []models.ChartVersion{{Version: "0.0.1", Digest: "123"}}},
+			{ID: "other-repo/my-chart", ChartVersions: []models.ChartVersion{{Version: "0.0.1", Digest: "123"}}},
+		}, []*models.Chart{
+			{ID: "my-repo/my-chart", ChartVersions: []models.ChartVersion{{Version: "0.0.1", Digest: "123"}}},
```
Comment (prydonius) [inline cmd/chartsvc/handler_test.go]: I think one other case here is if two different charts have an older version that is duplicated, but the latest versions are different (might be a strange edge case), but I think the expected thing to do in this case would be to not filter it out. wdyt? worth adding that test?

File: cmd/chartsvc/handler.go [hunk 5]
```diff
@@ -217,11 +217,16 @@ func newChartResponse(c *models.Chart) *apiResponse {
 }
 
 func newChartListResponse(charts []*models.Chart) apiListResponse {
+	// We will keep track of unique digest:chart to avoid duplicates
+	chartDigests := map[string]bool{}
 	cl := apiListResponse{}
 	for _, c := range charts {
-		cl = append(cl, newChartResponse(c))
+		digest := c.ChartVersions[0].Digest
+		if _, ok := chartDigests[digest]; !ok {
```
Comment (prydonius) [inline cmd/chartsvc/handler.go]: I still think it's useful here to add a comment like "filter out the chart if we've seen the same digest before" just as an added clarification to the comment above

File: cmd/chartsvc/handler_test.go [hunk 6]
```diff
@@ -126,28 +126,47 @@ func Test_newChartResponse(t *testing.T) {
...
+		}},
+		{"has a duplicated chart but with different versions", []*models.Chart{
```
Comment (prydonius) [inline cmd/chartsvc/handler_test.go]:
```suggestion
		{"has a duplicated older chart version but different latest versions", []*models.Chart{
```

File: cmd/chartsvc/handler.go (hunk 5 above)
Comment (andresmgot) [inline cmd/chartsvc/handler.go]:
```suggestion
		// Filter out the chart if we've seen the same digest before
		if _, ok := chartDigests[digest]; !ok {
```
